src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -56,13 +56,65 @@ static inline int MPIDI_OFI_win_progress_fence(MPIR_Win * win)
 
     r = MPIDI_OFI_WIN(win).syncQ;
 
-    while (r) {
-        MPIDI_OFI_win_request_t *next = r->next;
-        MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
-        r = next;
+    /* Should not free the RMA request when do_free = 0; manual RMA flush could happen in
+     * the middle of issuing one MPI-level RMA op and it could lead to premature freeing of
+     * the request. */
+    if (do_free) {
+        while (r) {
+            MPIDI_OFI_win_request_t *next = r->next;
+            MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
+            r = next;
+        }
+
+        MPIDI_OFI_WIN(win).syncQ = NULL;
+        MPIDI_OFI_WIN(win).counter = 1;
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_FLUSH);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_OFI_win_counter_check
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_OFI_win_counter_check(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    if (!MPIDI_OFI_ENABLE_DATA_AUTO_PROGRESS) {
+        if (MPIDI_OFI_WIN(win).counter % MPIR_CVAR_CH4_OFI_RMA_FLUSH_INTERVAL == 0) {
+            MPIDI_OFI_win_flush(win, MPIDI_OFI_KEEP_RMA_REQS);
+            MPIDI_OFI_WIN(win).counter = 1;
+        }
```

Comment:
If you already reset `counter` in `MPIDI_OFI_win_flush `, you do not reset it again here.

Reply:
I am actually not sure if we really need a "blocking wait" for the manual progress. Can we change it to nonblocking (i.e., you do not have to wait till `tcount == donecount` in `MPIDI_OFI_win_flush `)?

Reply:
> If you already reset `counter` in `MPIDI_OFI_win_flush `, you do not reset it again here.

This is a good point actually. I think the reset in `MPIDI_OFI_win_flush` should be removed and only this one should stay (because the one in `MPIDI_OFI_win_flush` is only triggered when the request can be freed).

Reply:
> I am actually not sure if we really need a "blocking wait" for the manual progress. Can we change it to nonblocking (i.e., you do not have to wait till `tcount == donecount` in `MPIDI_OFI_win_flush `)?

I think the problem is one of branding. This function shouldn't be named "flush" because it is actually intended to be a blocking progress function. It's only being used by fence operations. I've renamed it to `MPIDI_OFI_win_progress_fence_impl`.